source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+            result.flavor = Flavor::Constant;
+            result.value = value;
+            return result;
+        }
+
+        // We also need to be able to test if two lattice
+        // values are equal, so that we can avoid updating
+        // downstream dependencies if our knowledge about
+        // an instruction hasn't actually changed.
+        //
+        bool operator==(LatticeVal const& that)
+        {
+            return this->flavor == that.flavor
+                && this->value == that.value;
+        }
+
+        bool operator!=(LatticeVal const& that)
+        {
+            return !( *this == that );
+        }
+    };
+
+    // If we imagine a variable (actually an SSA phi node...) that
+    // might be assigned lattice value A at one point in the code,
+    // and lattice value B at another point, we need a way to
+    // combine these to form our knowledge of the possible value(s)
+    // for the variable.
+    //
+    // In terms of computation on a lattice, we want the "meet"
+    // operation, which computes the lower bound on what we know.
+    // If we interpret our lattice values as sets, then we are
+    // trying to compute the union.
+    //
+    LatticeVal meet(LatticeVal const& left, LatticeVal const& right)
+    {
+        // If either value is `None` (the empty set), then the union
+        // will be the other value.
+        //
+        if(left.flavor == LatticeVal::Flavor::None) return right;
+        if(right.flavor == LatticeVal::Flavor::None) return left;
+
+        // If either value is `Any` (the universal set), then
+        // the union is also the universal set.
+        //
+        if(left.flavor == LatticeVal::Flavor::Any) return LatticeVal::getAny();
+        if(right.flavor == LatticeVal::Flavor::Any) return LatticeVal::getAny();
+
+        // At this point we've ruled out the case where either value
+        // is `None` *or* `Any`, so we can assume both values are
+        // `Constant`s.
+        SLANG_ASSERT(left.flavor == LatticeVal::Flavor::Constant);
+        //
+        SLANG_ASSERT(right.flavor == LatticeVal::Flavor::Constant);
+
+        // If the two lattice values represent the *same* constant value
+        // (they are the same singleton set) then the union is that
+        // singleton set as well.
+        //
+        // TODO: This comparison assumes that constants with
+        // the same value with be represented with the
+        // same instruction, which is not *always*
+        // guaranteed in the IR today.
+        //
+        if(left.value == right.value)
+            return left;
+
+        // Otherwise, we have two distinct singleton sets, and their
+        // union should be a set with two elements. We can't represent
+        // that on the lattice for SCCP, so the proper lower bound
+        // is the universal set (`Any`)
+        //
+        return LatticeVal::getAny();
+    }
+
+    // During the execution of the SCCP algorithm, we will track our best
```

Comment:
I sometimes wonder if we had an integral associated with an instruction (perhaps reusable depending on context), such that it could be used to do fast mapping (like in serialization!), and other purposes. But then it makes the inst bigger, so there's that to balance too.
Here it could be used to remove the need for the DictionaryMap, and data stored directly (if it's a ptr sized int say) or a lookup into an array. Anyhoo...

Reply:
I've worked in a compiler that had such a thing. It could reduce some of the hashing that goes on, but at the cost of making certain code a bit more fiddly (since you are using a generic field to store different very specific data at different times), and potentially messing with the ability to do any analysis passes in parallel (not that there is a ton of opportunity there since as soon as you want to *modify* the code you need exclusive ownership anyway).

The current design definitely favors simplicity of the code over runtime efficiency, and that is a conscious choice to try to minimize the chance of bugs sneaking in when the code has lots of clever tricks for efficiency. The right solution IMHO is to beef up our testing so that we can make the representations more efficient in later changes with confidence.